### DISCUSSION TITLE
How to use ESP-NOW in zephyr?

### DISCUSSION THREAD

**Discussion by NineDayCC** [zephyrproject-rtos/zephyr#61254] [Q&A]
### Discussed in https://github.com/zephyrproject-rtos/zephyr/discussions/61055

<div type='discussions-op-text'>

<sup>Originally posted by **NineDayCC** August  2, 2023</sup>
Does zephyr support ESP-NOW? I found ESP-NOW deivers in _hal_espressif_, but I didn't find any APIs or Kconfig options to enable ESP-NOW.  
Is there any way to use ESP-NOW in zephyrOS?</div>

**Comment by ndrs-pst** [comment]
Hi, @NineDayCC.
After reviewing the `soc/xtensa/espressif_esp32/esp32/default.ld` in [Zephyr repository](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/xtensa/espressif_esp32/esp32/default.ld), I'm concerned that Zephyr does not yet support ESP-NOW.

This is because I couldn't find `libespnow.a`, which is located in the [hal_espressif repository](https://github.com/zephyrproject-rtos/hal_espressif/tree/zephyr/zephyr/blobs/lib/esp32), referenced in the linker script.
This omission suggests that ESP-NOW might not be integrated into Zephyr for the ESP32 platform.

Another aspect we need to consider is whether `libespnow.a` is independent of the ESP-IDF platform or not. 

ping @sylvioalves 👀

**Reply by rhvarrier** [reply]
@sylvioalves is there any plan to add support for esp now on zephyr?

**Comment by bhuvanchandradv** [comment]
Hi all,

Following up on this discussion — I've submitted a PR that adds a working ESP-NOW sample for Espressif ESP32 boards:

PR #108332

The sample (samples/boards/espressif/espnow) demonstrates:                                                               
- WiFi STA init + fixed-channel lock (no AP required)
- esp_now_init() with broadcast peer registration
- RX callback with RSSI, source MAC, and decoded heartbeat payload
- Beacon thread broadcasting a 14-byte heartbeat at a configurable interval
- Build-time role selection: SENDER / RECEIVER / BOTH (default : useful for two-board loopback)

Tested on esp32_devkitc/esp32/procpu (WROOM) with two devices talking to each other. Builds clean for esp32, esp32s3, and esp32c6 targets via Twister.

One thing worth noting for anyone working on this area: CONFIG_WIFI_ESP32 currently Kconfig-selects NET_L2_WIFI_MGMT -> NETWORKING -> NET_IPV4, which drags in the IP stack even though ESP-NOW never touches it. The sample works around this by minimising net buffer pool sizes. This is a known structural limitation in the Espressif WiFi Kconfig; a proper fix would need to decouple NET_L2_WIFI_MGMT selection from the raw WiFi blob init path.

Happy to address review feedback. If there's an existing tracking issue for ESP-NOW support I should link to, please let me know.

**Reply by NotPubi** [reply]
I just learned about zephyr and want to port my codebase but i need espnow, thanks for doing this. I don't understand what you said the problem and solution is, so hopefully it works out.
